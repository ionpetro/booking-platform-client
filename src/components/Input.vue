<template>
  <div class="form-group">
    <input
      :value="value"
      :type="customType"
      class="text-input"
      :placeholder="placeholder"
      :class="errors ? 'input-errors' : ''"
      @input="$emit('input', $event.target.value)"
      @focus="$emit('focused', labelName)"
    />
    <label class="text-label" :class="errors ? 'label-errors' : ''" for="name">{{
      labelName
    }}</label>
    <div class="icon" @click="showPassword" v-if="labelName === 'Password'">
      <img v-show="showEye" src="@/assets/icons/Eye.svg" alt="" />
      <img v-show="!showEye" src="@/assets/icons/Eye-Off.svg" alt="" />
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      showEye: true,
      customType: this.type
    };
  },
  props: {
    value: {
      required: true,
      type: String
    },
    name: {
      required: true,
      type: String
    },
    labelName: {
      required: true,
      type: String
    },
    errors: {
      required: false,
      type: [String || null]
    },
    placeholder: {
      required: true,
      type: String
    },
    type: {
      required: true,
      default: 'text',
      type: String
    }
  },
  methods: {
    // toggle password input
    showPassword() {
      if (this.name === 'password') {
        if (this.customType === 'password') {
          this.customType = 'text';
          this.showEye = !this.showEye;
        } else {
          this.customType = 'password';
          this.showEye = !this.showEye;
        }
      }
    }
  }
};
</script>

<style lang="scss">
.form-group {
  position: relative;
}
// text and password input
.text-input {
  background: #ffffff;

  /* Gray/ dim */
  border: 1px solid $gray-dim;
  box-sizing: border-box;
  height: 56px;
  width: 480px;
  padding-left: 1.5em;
  border-radius: 20px;
  color: $secondary;
  // font-size: $base-font-size;
  font-size: 1rem;
  margin-top: 55px;

  &:focus {
    border: 1px solid $secondary;
    outline: none;
    & + label {
      color: $secondary !important;
    }
  }
}
// text and password label
.text-label {
  color: $gray-dim;
  position: absolute;
  z-index: 1;
  left: 17px;
  top: 45px;
  padding: 0 8px;
  background-color: white;
  font-size: 0.8rem;
}

.input-errors {
  border: 1px solid $red-medium;
}

.label-errors {
  color: $red-medium;
}

//checkbox input
[type='checkbox']:not:checked,
[type='checkbox']:checked {
  border: 2px solid black;
}

.icon {
  position: absolute;
  // input width - (icon width + distance from edge)
  left: #{480- (24+25)}px;
  bottom: 10px;
}
</style>
