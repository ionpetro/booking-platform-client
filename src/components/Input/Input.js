const types = {
  text: 'text',
  password: 'password'
};

export default {
  data() {
    return {
      showEye: true,
      customType: this.type
    };
  },
  props: {
    value: String,
    name: String,
    labelName: String,
    placeholder: String,
    errors: {
      required: false,
      type: [String || null]
    },
    type: {
      required: true,
      default: types.text,
      type: String,
      validator: value => Object.values(types).includes(value)
    }
  },
  methods: {
    // toggle password input
    showPassword() {
      if (this.name !== types.password) {
        return;
      }
      this.customType = this.customType === types.password ? types.text : types.password;
      this.showEye = !this.showEye;
    }
  }
};
