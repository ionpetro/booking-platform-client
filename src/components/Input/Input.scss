.form-group {
  position: relative;
}
// text and password input
.text-input {
  background: #ffffff;

  /* Gray/ dim */
  border: 1px solid $gray-dim;
  box-sizing: border-box;
  height: 56px;
  width: 480px;
  padding-left: 1.5em;
  border-radius: 20px;
  color: $secondary;
  // font-size: $base-font-size;
  font-size: 1rem;
  margin-top: 55px;

  &:focus {
    border: 1px solid $secondary;
    outline: none;
    & + label {
      color: $secondary !important;
    }
  }
}
// text and password label
.text-label {
  color: $gray-dim;
  position: absolute;
  z-index: 1;
  left: 17px;
  top: 45px;
  padding: 0 8px;
  background-color: white;
  font-size: 0.8rem;
}

.input-errors {
  border: 1px solid $red-medium;
}

.label-errors {
  color: $red-medium;
}

//checkbox input
[type='checkbox']:not:checked,
[type='checkbox']:checked {
  border: 2px solid black;
}

.icon {
  position: absolute;
  // input width - (icon width + distance from edge)
  left: #{480- (24+25)}px;
  bottom: 10px;
}
