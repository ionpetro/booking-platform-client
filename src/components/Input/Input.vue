<template>
  <div class="input">
    <input
      :value="value"
      :type="customType"
      :placeholder="placeholder"
      :class="['input__text', {input__errors: errors}]"
      @input="$emit('input', $event.target.value)"
      @focus="$emit('focused')"
    />
    <label class="label__text" :class="['label__text', {label__errors: errors}]" for="name">{{
      labelName
    }}</label>
    <div class="input__icon" @click="showPassword" v-if="type === 'password'">
      <img v-show="showEye" src="@/assets/icons/Eye.svg" alt="show password" />
      <img v-show="!showEye" src="@/assets/icons/Eye-Off.svg" alt="hide password" />
    </div>
  </div>
</template>

<script src="./Input.js"></script>

<style src="./Input.scss" lang="scss"></style>
