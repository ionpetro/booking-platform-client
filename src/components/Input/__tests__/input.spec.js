import {shallowMount} from '@vue/test-utils';
import Input from '../Input.vue';

describe('Input', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Input, {
      propsData: {
        type: 'text',
        name: 'password'
      }
    });
  });

  describe('when loaded', () => {
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('shows password on icon click', async () => {
      wrapper = shallowMount(Input, {
        propsData: {
          type: 'password',
          name: 'password'
        }
      });
      expect(wrapper.vm.$data.customType).toEqual('password');
      expect(wrapper.vm.$data.showEye).toBeTruthy();

      await wrapper.find('.input__icon').trigger('click');
      expect(wrapper.vm.$data.customType).toEqual('text');
      expect(wrapper.vm.$data.showEye).toBeFalsy();
    });

    it('hides password on icon click', async () => {
      wrapper = shallowMount(Input, {
        propsData: {
          type: 'password',
          name: 'password'
        }
      });

      // turn password field to text
      wrapper.vm.$data.customType = 'text';
      wrapper.vm.$data.showEye = false;

      await wrapper.find('.input__icon').trigger('click');
      expect(wrapper.vm.$data.customType).toEqual('password');
      expect(wrapper.vm.$data.showEye).toBeTruthy();
    });
    it('does not change input type when its text input', () => {
      wrapper = shallowMount(Input, {
        propsData: {
          type: 'text',
          name: 'text'
        }
      });
      wrapper.vm.showPassword();
      expect(wrapper.vm.$data.customType).toEqual('text');
    });
  });
});
