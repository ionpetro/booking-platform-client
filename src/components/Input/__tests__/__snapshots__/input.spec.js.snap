// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input when loaded renders correctly 1`] = `
<div
  class="input"
>
  <input
    class="input__text"
    type="text"
  />
   
  <label
    class="label__text label__text"
    for="name"
  >
    
  </label>
   
  <!---->
</div>
`;
